### PR TITLE
[Workflow] Virtual actor writer - Part 2

### LINKED ISSUE #16989: virtual actor support - write

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Second part of virtual actor implementation.

The key idea of the implementation:

1. Now every step execution returns 2 objects, instead of 1. The two objects are captured in `WorkflowExecutionResult`. The two objects are state and output. State is the object to persist and pass down to later steps; output is volatile and returns to the user.
2. We define an `WorkflowRef` object to dynamically form an edge in the DAG in the graph. This is very useful for virtual actor writers, because we can create an edge to the previous virtual actor writer step, so it will be automatically queued after the previous step and depends on the previous state. 

## Related issue number

Closes #16989 
Closes #17279. (Fail to cherrypick due to merging conflicts. The commit to cherrypick is a 3-way merging node.)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by suquark** [comment]
Still need to update some tests, but already ready for review

File: python/ray/experimental/workflow/workflow_storage.py [hunk 1]
```diff
@@ -236,9 +242,11 @@ async def _inspect_step(self, step_id: StepID) -> StepInspectResult:
                 self._workflow_id, step_id)
             input_object_refs = metadata["object_refs"]
             input_workflows = metadata["workflows"]
+            input_workflow_refs = metadata["workflow_refs"]
             max_retries = metadata.get("max_retries")
             catch_exceptions = metadata.get("catch_exceptions")
             ray_options = metadata.get("ray_options", {})
+            step_type = StepType[metadata.get("step_type")]
         except storage.DataLoadError:
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_storage.py:250]: This part:
```
try:
   ...
except xxx:
   return 
return 
```
Is a little bit verbose. How about this:

```
try:
  metadata = ...
  return StepInspectResult(
      object_refs=metadata["object_refs"]
     .....
   )
except ...
   return xxx
```

File: python/ray/experimental/workflow/workflow_access.py [hunk 2]
```diff
@@ -106,44 +116,77 @@ class WorkflowManagementActor:
...
+        except Exception:
+            return None
+
+    def run_or_resume(self, workflow_id: str, ignore_existing: bool = False
+                      ) -> "WorkflowExecutionResult":
+        """Run or resume a workflow.
+
+        Args:
+            workflow_id: The ID of the workflow.
+            ignore_existing: Ignore we already have an existing output.
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_access.py:152]: When set false, raise an exception if there has already been a workflow with this id running.

File: python/ray/experimental/workflow/virtual_actor_class.py [hunk 3]
```diff
@@ -380,32 +378,47 @@ def __getattr__(self, item):
...
-        flattened_args = self._metadata.flatten_args(method_name, args, kwargs)
         cls = self._metadata.cls
         method = getattr(cls, method_name, None)
         if method is None:
             raise AttributeError(f"Method '{method_name}' does not exist.")
+        readonly = getattr(method, "__virtual_actor_readonly__", False)
+
+        flattened_args = self._metadata.flatten_args(method_name, args, kwargs)
+        if not readonly:
+            if method_name == "__init__":
```
Comment (iycheng) [inline python/ray/experimental/workflow/virtual_actor_class.py:389]: Could we have one test case to cover the failure of `__init__`?

File: python/ray/experimental/workflow/workflow_access.py [hunk 4]
```diff
@@ -106,44 +116,77 @@ class WorkflowManagementActor:
 
     def __init__(self, store: "storage.Storage"):
         self._store = store
-        self._workflow_outputs: Dict[str, ray.ObjectRef] = {}
+        self._workflow_outputs: Dict[str, LatestWorkflowOutput] = {}
+        # Cache step output.
+        self._step_output_cache: Dict[str, LatestWorkflowOutput] = {}
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_access.py:121]: Please add some comments about why we need these two. I feel we can easily mix the purpose of these two things.

File: python/ray/experimental/workflow/workflow_access.py [hunk 5]
```diff
@@ -214,33 +257,36 @@ def actor_ready(self, actor_id: str) -> ray.ObjectRef:
...
+        latest_output = LatestWorkflowOutput(result.state, workflow_id,
+                                             step_id)
+        self._workflow_outputs[workflow_id] = latest_output
         wf_store = workflow_storage.WorkflowStorage(workflow_id, self._store)
         wf_store.save_workflow_meta(
             common.WorkflowMetaData(common.WorkflowStatus.RUNNING))
-        self._step_status[workflow_id] = {}
-        return output
+        self._step_status.setdefault(workflow_id, {})
+        return result.state
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_access.py:289]: The question here: if it's a normal workflow, will it be wrong? should we return `result.output`? Does this means here `result.state == result.output`?

File: python/ray/experimental/workflow/step_executor.py [hunk 6]
```diff
@@ -283,53 +340,24 @@ def _workflow_step_executor(
...
-        step_inputs,
-        outer_most_step_id,
-        workflow_data.catch_exceptions,
-        workflow_data.max_retries,
-        last_step_of_workflow=True)
-    if readonly:
-        return ret[1]  # only return output. skip state
-    return ret
+@dataclass
+class _BakedWorkflowInputs:
```
Comment (iycheng) [inline python/ray/experimental/workflow/step_executor.py:344]: Please give some comments about that's this for.

**Comment by iycheng** [comment]
Overall I feel fine about this PR. Due to some issues, there are a lot of changes everywhere. Could you give more comment about the new concept you introduced so that it can help people understand what are they and why do we need them, especially WorkflowRef and _BakedWorkflowInputs which are new concepts here and not straightforward to understand.

File: python/ray/experimental/workflow/workflow_access.py (hunk 5 above)
Comment (suquark) [reply python/ray/experimental/workflow/workflow_access.py:289]: state is for both workflow and virtual actor writer (state is the object to persist)

File: python/ray/experimental/workflow/common.py [hunk 7]
```diff
@@ -59,14 +87,29 @@ def to_metadata(self) -> Dict[str, Any]:
...
             "catch_exceptions": self.catch_exceptions,
             "ray_options": self.ray_options,
         }
 
 
+@dataclass
+class WorkflowExecutionResult:
+    """Dataclass for holding workflow execution result."""
+    # Part of result to persist in a storage and pass to the next step.
+    persisted_output: "ObjectRef"
```
Comment (iycheng) [inline python/ray/experimental/workflow/common.py:104]: Thanks!

File: python/ray/experimental/workflow/step_executor.py [hunk 8]
```diff
@@ -39,8 +43,51 @@ def _resolve_object_ref(ref: ObjectRef) -> Tuple[Any, ObjectRef]:
...
+            try:
+                output, _ = _resolve_object_ref(step_ref)
+                get_cached_step = True
+            except Exception:
+                get_cached_step = False
+        if not get_cached_step:
+            wf_store = workflow_storage.get_workflow_storage()
+            try:
+                output = wf_store.load_step_output(workflow_ref.step_id)
+            except Exception:
```
Comment (iycheng) [inline python/ray/experimental/workflow/step_executor.py:76]: Should it be DataLoadException?

Comment (iycheng) [reply python/ray/experimental/workflow/step_executor.py:76]: What if the step doesn't exist due to some reason?

**Review by fishbone — COMMENTED** [review]
Thanks for the work here! `WorkflowRef` is a nice concept to have which can be used to retrieve results from steps in the future.

When I'm reading the code, I feel some parts them is a little bit hard to understand due to the organization. For example the code and the def are not close, usually this is because the module is not organized well.  I assigned an issue to you https://github.com/ray-project/ray/issues/17445 .

We can talk about this when ft testing is finished.

**Review by fishbone — APPROVED** [review]

**Comment by suquark** [comment]
Failed CI tests seem not related.

**Pull request merged**